src/analyzer/analyzerthread.cpp

```diff
@@ -0,0 +1,326 @@
+    }
+
+    mixxx::AudioSource::OpenParams openParams;
+    openParams.setChannelCount(kAnalysisChannels);
+
+    while (waitUntilWorkItemsFetched()) {
+        DEBUG_ASSERT(m_currentTrack);
+        kLogger.debug() << "Analyzing" << m_currentTrack->getLocation();
+
+        // Get the audio
+        const auto audioSource =
+                SoundSourceProxy(m_currentTrack).openAudioSource(openParams);
+        if (!audioSource) {
+            kLogger.warning()
+                    << "Failed to open file for analyzing:"
+                    << m_currentTrack->getLocation();
+            emitDoneProgress(kAnalyzerProgressUnknown);
+            continue;
+        }
+
+        bool processTrack = false;
+        for (auto const& analyzer: m_analyzers) {
+            // Make sure not to short-circuit initialize(...)
+            if (analyzer->initialize(
+                    m_currentTrack,
+                    audioSource->sampleRate(),
+                    audioSource->frameLength() * kAnalysisChannels)) {
+                processTrack = true;
+            }
+        }
+
+        if (processTrack) {
+            const auto analysisResult = analyzeAudioSource(audioSource);
+            DEBUG_ASSERT(analysisResult != AnalysisResult::Pending);
+            if ((analysisResult == AnalysisResult::Complete) ||
+                    (analysisResult == AnalysisResult::Partial)) {
+                // The analysis has been finished, and is either complete without
+                // any errors or partial if it has been aborted due to a corrupt
+                // audio file. In both cases don't reanalyze tracks during this
+                // session. A partial analysis would otherwise be repeated again
+                // and again, because it is very unlikely that the error vanishes
+                // suddenly.
+                emitBusyProgress(kAnalyzerProgressFinalizing);
+                // This takes around 3 sec on a Atom Netbook
+                for (auto const& analyzer: m_analyzers) {
+                    analyzer->finalize(m_currentTrack);
+                }
+                emitDoneProgress(kAnalyzerProgressDone);
+            } else {
+                for (auto const& analyzer: m_analyzers) {
+                    analyzer->cleanup(m_currentTrack);
+                }
+                emitDoneProgress(kAnalyzerProgressUnknown);
+            }
+        } else {
+            kLogger.debug() << "Skipping track analysis because no analyzer initialized.";
+            emitDoneProgress(kAnalyzerProgressDone);
+        }
+    }
+    DEBUG_ASSERT(isStopping());
+
+    m_analyzers.clear();
+
+    emitProgress(AnalyzerThreadState::Exit);
+}
+
+void AnalyzerThread::submitNextTrack(TrackPointer nextTrack) {
+    DEBUG_ASSERT(!m_nextTrack.getValue());
+    m_nextTrack.setValue(std::move(nextTrack));
+}
+
+WorkerThread::FetchWorkResult AnalyzerThread::tryFetchWorkItems() {
+    DEBUG_ASSERT(!m_currentTrack);
+    m_currentTrack = m_nextTrack.getValueOnce();
```

Comment:
we should only use the move semantic if we do not access the variable afterwards. This is not the case here, because it is a member variable. I also think it is dangerous to put a shared pointer into a ring buffer. 
Cant we find a workaround eg using an Atomic pointer to the shared pointer.